### ISSUE TITLE
Is this DALEX.predict_parts for SHAP support distributed computation?

### ISSUE THREAD

**Issue by MichaelChaoLi-cpu** [ModelOriented/DALEX#527]
In one machine learning research, we hope to obtain the SHAP values of all observations. However, it is too time-consuming, even impossible. We do have HPC. 

We also find that the DALEX.predict_parts can be parallelized. We try to use dask.distribute to further increase speed but nothing changed.

```
import os
import dask_mpi as dm
from dask.distributed import Client, progress, LocalCluster

from datetime import datetime

import joblib
from dask.distributed import Client, progress

from sklearn.datasets import make_regression
from sklearn.ensemble import RandomForestRegressor

dm.initialize(local_directory=os.getcwd(),  nthreads=36)
X, y = make_regression(n_samples = 500000, n_features = 60, random_state=1)

client = Client()
model = RandomForestRegressor(n_estimators=1000, oob_score=True, random_state=1)

with joblib.parallel_backend("dask"): model.fit(X, y)

import dalex as dx

model_14feature_rf_exp = dx.Explainer(model, X, y,  
                                      label = "RF Pipeline")
test_obs = X.iloc[0:1,:]

start = datetime.now()
shap_test = model_14feature_rf_exp.predict_parts(test_obs, type = 'shap')
end = datetime.now()
print(f"No Dask Time taken: {end - start}")

start = datetime.now()
with joblib.parallel_backend("dask"): shap_test = model_14feature_rf_exp.predict_parts(test_obs, type = 'shap')
end = datetime.now()
print(f"With Dask Time taken: {end - start}")
```

If dask works, With Dask Time taken should be smaller that  No Dask Time taken, since the second computation use 4 nodes of NPC, while the first calculation only use one node.

In a nutshell, Is this DALEX.predict_parts for SHAP support distributed computation of 'dask' or something else?

Thank you.

**Comment by hbaniecki** [comment]
Hi, `predict_parts` has the following useful parameters:
* `N`, which is the number of observations that will be sampled from the data attribute before the calculation of variable attributions. The default is `None` which means all data. You can set it to 200, 500, or 1000 for example.
* `B`, which is the number of random paths to calculate variable attributions (default is `25`). You can decrease this parameter for faster approximation at the cost of estimation error.
*  `processes=1`, which is the number of parallel processes to use in calculations: iterated over B (default is 1, which means no parallel computation). Does it help in your case?

https://dalex.drwhy.ai/python/api/index.html#dalex.Explainer.predict_parts

**Comment by MichaelChaoLi-cpu** [comment]
Hi, dear author, @hbaniecki 

Thanks for your reply. I read this document. ```processes=1``` is a kind of parallel, but not effective. E.g. my HPC has 144 cores but I only need ```B=25``` at most. In other words, most cores are idle.

I want to know whether there is a way to set something like ```n_jobs``` in the ```sklearn.ensemble.RandomForestRegressor```. 

There is another problem: on windows, all the cores are used when I run ```predict_parts```, since ```RandomForestRegressor(n_estimators=1000, oob_score=True, random_state=1, n_jobs = -1)```. But, when I go to Linux, the story change, even though I have more cores in a single machine, it still takes a longer time. 
I have no idea about this. 

Thank you.

**Comment by MichaelChaoLi-cpu** [comment]
Hi, @hbaniecki 

Supplementary information:
```
/home/usr6/q70176a/anaconda3/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(

Preparation of a new explainer is initiated

  -> data              : 89273 rows 49 cols
  -> target variable   : Parameter 'y' was a pandas.DataFrame. Converted to a numpy.ndarray.
  -> target variable   : 89273 values
  -> model_class       : sklearn.ensemble._forest.RandomForestRegressor (default)
  -> label             : RF Pipeline
  -> predict function  : <function yhat_default at 0x2b4c59cb9c10> will be used (default)
  -> predict function  : Accepts pandas.DataFrame and numpy.ndarray.
  -> predicted values  : min = 1.38, mean = 24.3, max = 36.0
  -> model type        : regression will be used (default)
  -> residual function : difference between y and yhat (default)
  -> residuals         : min = -11.3, mean = 0.00982, max = 7.26
  -> model_info        : package sklearn
```

This is the return of ```model_14feature_rf_exp = dx.Explainer(model, X, y,  label = "RF Pipeline")```.
I do not know whether this warning ```/home/usr6/q70176a/anaconda3/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(
 ``` influences the speed of ```model.predict()```.
 
Best

**Comment by hbaniecki** [comment]
Hi @MichaelChaoLi-cpu, Your goal is "to obtain the SHAP values of all observations". 

How about parallelizing the calculations by iterating over the list of observations? If you have 144 cores, you can compute 144x `predict_parts` explanations at once using python `multiprocessing.Pool` https://docs.python.org/3/library/multiprocessing.html. Also, as I said, it should be considerably faster to estimate Shapley values using `N=1000` observations (or less) than when trying to use the whole dataset.

Next, from my experience, `RandomForestRegressor` is very slow: both in training and inference. For faster explanation estimation, we need the lowest inference overhead possible. Have you considered using a random forest implementation from the `lightgbm` package? Apparently, it also works with Dask https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.DaskLGBMRegressor.html.

Finally, when working with tree-ensembles, consider using the TreeSHAP algorithm implemented in the `shap` Python package.

```python
#%%
from sklearn.datasets import make_regression

X, y = make_regression(n_samples=5000, n_features=50, random_state=1)

#%%
from lightgbm import LGBMRegressor

model = LGBMRegressor(
    boosting_type="rf", 
    n_estimators=1000, 
    random_state=1,
    # make the forest actually random:
    subsample_freq=1, 
    subsample=0.8**2,
    colsample_bytree=0.8,
    feature_fraction_bynode=0.8
)
model.fit(X, y)

#%%
import dalex as dx

explainer = dx.Explainer(model, X, y)

#%%
shap = explainer.predict_parts(X[[1]], type="shap", N=1000, B=20)

#%%
from multiprocessing import Pool

def parallel_shap(observation_id):
    print(observation_id)
    shap = explainer.predict_parts(X[[observation_id]], type="shap", N=1000, B=20)
    return shap.result

#%%
# for real use-case: first save model/explainer and then load it in each process
if __name__=='__main__':
    pool = Pool(processes=2)
    observation_ids = list(range(4))
    result = pool.map(parallel_shap, observation_ids)
    pool.close()
    print(result)
```

**Comment by hbaniecki** [comment]
assuming it's fixed
